plugin/updater_oneapi/regression_obj_oneapi.cc

```diff
@@ -0,0 +1,180 @@
+
+    // TODO: add label_correct check
+    label_correct_.Resize(1);
+    label_correct_.Fill(1);
+
+    cl::sycl::buffer<bst_float, 1> preds_buf(preds.HostPointer(), preds.Size());
+    cl::sycl::buffer<bst_float, 1> labels_buf(info.labels_.HostPointer(), info.labels_.Size());
+    cl::sycl::buffer<GradientPair, 1> out_gpair_buf(out_gpair->HostPointer(), out_gpair->Size());
+
+    bool is_null_weight = info.weights_.Size() == 0;
+    auto scale_pos_weight = param_.scale_pos_weight;
+    if (!is_null_weight) {
+      CHECK_EQ(info.weights_.Size(), ndata)
+          << "Number of weights should be equal to number of data points.";
+
+      cl::sycl::buffer<bst_float, 1> weights_buf(info.weights_.HostPointer(), info.weights_.Size());
+
+      qu_.submit([&](cl::sycl::handler& cgh) {
+        auto preds_acc     = preds_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto labels_acc    = labels_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto weights_acc   = weights_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto out_gpair_acc = out_gpair_buf.get_access<cl::sycl::access::mode::write>(cgh);
+        cgh.parallel_for<class GetGradientWeights>(cl::sycl::range<1>(ndata), [=](cl::sycl::id<1> pid) {
+          int idx = pid[0];
+          bst_float p = Loss::PredTransform(preds_acc[idx]);
+          bst_float w = weights_acc[idx];
+          bst_float label = labels_acc[idx];
+          if (label == 1.0f) {
+            w *= scale_pos_weight;
+          }
+          out_gpair_acc[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                            Loss::SecondOrderGradient(p, label) * w);
+        });
+      }).wait();
+    } else {
+      qu_.submit([&](cl::sycl::handler& cgh) {
+        auto preds_acc = preds_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto labels_acc = labels_buf.get_access<cl::sycl::access::mode::read>(cgh);
+        auto out_gpair_acc = out_gpair_buf.get_access<cl::sycl::access::mode::write>(cgh);
+        cgh.parallel_for<class GetGradientNoWeights>(cl::sycl::range<1>(ndata), [=](cl::sycl::id<1> pid) {
+          int idx = pid[0];
+          bst_float p = Loss::PredTransform(preds_acc[idx]);
+          bst_float w = 1.0f;
+          bst_float label = labels_acc[idx];
+          if (label == 1.0f) {
+            w *= scale_pos_weight;
+          }
+          out_gpair_acc[idx] = GradientPair(Loss::FirstOrderGradient(p, label) * w,
+                                            Loss::SecondOrderGradient(p, label) * w);
+        });
+      }).wait();
+    }
+  }
+
+ public:
+  const char* DefaultEvalMetric() const override {
+    return Loss::DefaultEvalMetric();
+  }
+
+  void PredTransform(HostDeviceVector<float> *io_preds) override {
+    size_t const ndata = io_preds->Size();
+
+    cl::sycl::buffer<bst_float, 1> io_preds_buf(io_preds->HostPointer(), io_preds->Size());
+
+    qu_.submit([&](cl::sycl::handler& cgh) {

```

Comment:
Is it possible to use a global queue and don't block on it?

Reply:
This will be one of our further steps, when we move to proper device dispatching (we are still in discussion on how to do it).
However, we need to find a good place to store the single queue for all parts of training flow and/or implement mechanisms to pass the queue between this parts (updater, gradient computation, obj. function, etc.).